### PR TITLE
Update integration test to include sample order comparison

### PR DESCRIPTION
**Pull request by karan6181**
## Description of changes:
- As part of testing streaming feature regularly to capture any regression, I had to add the below changes to make certain test working, such as.
  - Validating number of files locally
  - Validating the epoch iteration time
  - Validating the sample orders
- More details is in PR https://github.com/mosaicml/testing/pull/43

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: regression/iterate_data.py [hunk 1]
```diff
@@ -104,10 +109,9 @@ def main(args: Namespace, kwargs: dict[str, str]) -> None:
     iter_time = []
     start_time = 0
     for epoch in range(args.epochs):
-        if epoch > 0:
-            start_time = time.time()
+        start_time = time.time()
         for batch in dataloader:
-            if args.sample_order_file is not None:
+            if args.sample_order_file is not None and epoch == 0:
```
Comment (snarayan21) [inline regression/iterate_data.py:114]: curious why we're not writing samples to the file past epoch 0?

File: regression/iterate_data.py [hunk 2]
```diff
@@ -30,12 +34,13 @@ def parse_args() -> tuple[Namespace, dict[str, str]]:
...
                       help='Number of epochs to iterate over the data')
-    args.add_argument('--validate-files', default=False, action='store_true', help='Verify files')
-    args.add_argument('--sample-order-file', type=str, help='File location to store sample order')
-    args.add_argument('--validate-iter-time',
+    args.add_argument('--validate-num-files',
                       default=False,
                       action='store_true',
-                      help='Test iter time')
+                      help='Verify number of dataset files')
+    args.add_argument('--sample-order-file', type=str, help='File location to store sample order')
```
Comment (snarayan21) [inline regression/iterate_data.py:41]: when we automatically run this test, is there a default argument to set here or will CORGI handle that for us?

File: regression/utils.py [hunk 3]
```diff
@@ -204,3 +205,65 @@ def delete_oci(remote_dir: str) -> None:
 
     for filenames in objects.data.objects:
         oci_client.delete_object(namespace, obj.netloc, filenames.name)
+
+
+def compare_files(files: list[str], identical: bool = False) -> None:
```
Comment (snarayan21) [inline regression/utils.py:210]: can we use the library difflib for this? gets the diff between two files. that's what we need right? Example here:
[](https://www.tutorialspoint.com/How-to-find-difference-between-2-files-in-Python#:~:text=We%20invoke%20the%20open(),file%20names%20in%20the%20output.)

**Review by snarayan21 — APPROVED** [review]
lgtm, just had some clarifying questions

File: regression/utils.py (hunk 3 above)
Comment (karan6181) [reply regression/utils.py]: So the method `compare_files()` is not just the diff comparison, it is actually a sample order comparison between multiple files across all GPUs. For example, each line in an output file contains a sample id of `batch_size * n_gpus_across_all_nodes`  samples. For deterministic sample ordering, given same NCN and global batch_size, each gpu gets some sample ordering, so with one forward pass across all the gpus, the set of sample ids between any two runs should be same, however, the samples can be in different order (in different gpu or in different position in a batch size) in the same forward pass.

File: regression/iterate_data.py (hunk 2 above)
Comment (karan6181) [reply regression/iterate_data.py:42]: CORGI provides the file path.

File: regression/iterate_data.py (hunk 1 above)
Comment (karan6181) [reply regression/iterate_data.py:115]: Good point. I will have to change the logic in future for testing different sampling_method

**Pull request merged**